morpheus/io/data_record.py

```diff
@@ -0,0 +1,275 @@
+# SPDX-FileCopyrightText: Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import io
+from typing import Any
+from typing import Union
+
+import fsspec
+import pandas as pd
+import pyarrow.parquet as pq
+
+import cudf
+
+
+class DataRecord:
+    """Class for managing data records in different storage types and formats.
+
+    Attributes:
+        VALID_STORAGE_TYPES (tuple): Allowed storage types.
+        VALID_FILE_FORMATS (tuple): Allowed file formats.
+    """
+
+    VALID_STORAGE_TYPES = ('in_memory', 'filesystem')
+    VALID_FILE_FORMATS = ('parquet', 'csv')
+
+    def __init__(self,
+                 data_source: Union[io.BytesIO, str],
+                 data_label: str,
+                 storage_type: str,
+                 file_format: str,
+                 copy_from_source: bool = False):
+        """Initialize a DataRecord instance.
+
+        Args:
+            data_source (Union[io.BytesIO, str]): Data source, either a file path or a BytesIO object.
+            data_label (str): Label for the data record.
+            storage_type (str): Storage type, either 'in_memory' or 'filesystem'.
+            file_format (str): File format, either 'parquet' or 'csv'.
+            copy_from_source (bool, optional): If True, copy data from the source. Defaults to False.
+        """
+
+        self._copy_from_source = copy_from_source
+        self._data = None
+        self._data_label = data_label
+        self._file_format = file_format
+        self._fs = fsspec.filesystem('file')
+        self._num_rows = 0
+        self._owner = False
+        self._backing_source = None
+        self._storage_type = storage_type
+
+        if self._file_format not in self.VALID_FILE_FORMATS:
+            raise ValueError(f"Invalid file_format '{self._file_format}'")
+
+        if self._storage_type not in self.VALID_STORAGE_TYPES:
+            raise ValueError(f"Invalid storage_type '{self._storage_type}'")
+
+        if self._file_format == 'csv':
+            self._data_reader = cudf.read_csv
+        elif self._file_format == 'parquet':
+            self._data_reader = cudf.read_parquet
+
+        self._store(data_source)
+
+    def __del__(self):
+        """Delete the DataRecord instance."""
+        try:
+            if (self._owner and self._storage_type == 'filesystem'):
+                self._fs.rm(self._backing_source)
```

Comment:
Things like this lead me to believe that a better implementation would be one with `self._data` point to an abstract class `DataRecordStorage` which handles all of the actually reading/writing, deletion, etc. Then you have different implementations for in memory, filesystem, etc. And you could even make utility classes that implement the interface but act as a cache and fallback through chained implementations in case of a cache miss.

Keeping pairs of properties like this working correctly will be difficult as the number of storage types increases.